src/Native/System.Native/pal_mount.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_mount.h"
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+
+// Check if we should use getmntinfo or /proc/mounts
+#if HAVE_MNTINFO
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#include <mntent.h>
+#define STRING_BUFFER_SIZE 8192
+#endif
+
+#if HAVE_MNTINFO
+
+int GetMountInfo(MountPointFound onFound)
+{
+    // getmntinfo returns pointers to OS-internal structs, so we don't need to worry about free'ing the object
+    int result = 0;
+    struct statfs* mounts = nullptr;
+    int count = getmntinfo(&mounts, 0);
+    if (count > 0)
+    {
+        for (int32_t i = 0; i < count; i++)
+        {
+            int32_t nameLength = strlen(mounts[i].f_mntonname);
+            onFound(mounts[i].f_mntonname, nameLength);
+        }
+    }
+    else
+    {
+        // Something went wrong so set our result to errno
+        result = errno;
+    }
+    
+    return result;
+}
+
+#else
+
+int GetMountInfo(MountPointFound onFound)
+{
+    int result = 0;
+    FILE* fp = setmntent("/proc/mounts", MNTOPT_RO);
+    if (fp == nullptr)
+    {
+        result = errno;
+    }
+    else
+    {
+        // The _r version of getmntent needs all buffers to be passed in; however, we don't know how big of a string
+        // buffer we will need, so pick something that seems like it will be big enough.
```

Comment:
After investigating more, this seems extremely unlikely to happen; the mount point, name of the file system, and mount options would all have to add up to over 8k chars...very very unlikely. A fix for this would be tricky since the current implementation calls back into the C# code each time it comes across a mount point, so there would be duplicates passed across each iteration of increasing the buffer size. The callback implementation was chosen to prevent malloc'ing on the native side and requiring the managed side to free it, so there would need to be a redesign if we choose to handle this case. @stephentoub do you think we should handle this case or is the odds of a mount point having 8k worth of strings low enough that we can skip it?